src/crypto/public_key/elliptic/ecdsa.js

```diff
@@ -25,25 +25,52 @@
 
 'use strict';
 
+import BN from 'bn.js';
+import ASN1 from 'asn1.js';
+import jwkToPem from 'jwk-to-pem';
+
 import curves from './curves.js';
 import BigInteger from '../jsbn.js';
+import config from '../../../config';
+import enums from '../../../enums.js';
+import util from '../../../util.js';
+import base64 from '../../../encoding/base64.js';
+
+const webCrypto = util.getWebCrypto();
+const webCurves = curves.webCurves;
+const nodeCrypto = util.getNodeCrypto();
+const nodeCurves = curves.nodeCurves;
+
+var ECDSASignature = ASN1.define('ECDSASignature', function() {
+  this.seq().obj(
+    this.key('r').int(),  // FIXME int or BN?
+    this.key('s').int()   // FIXME int or BN?
+  );
+});
 
 /**
  * Sign a message using the provided key
  * @param  {String}      oid        Elliptic curve for the key
- * @param  {BigInteger}  Q          Public key used to sign
  * @param  {enums.hash}  hash_algo  Hash algorithm used to sign
  * @param  {Uint8Array}  m          Message to sign
-* @param  {BigInteger}  d          Private key used to sign
+ * @param  {BigInteger}  d          Private key used to sign
  * @return {{r: BigInteger, s: BigInteger}}  Signature of the message
  */
-function sign(oid, hash_algo, m, d) {
+async function sign(oid, hash_algo, m, d) {
+  var signature;
   const curve = curves.get(oid);
+  hash_algo = hash_algo ? hash_algo : curve.hash;
```

Comment:
So, when signing a message, it turns out that hash_algo just comes directly from the config's preferred hash algorithm (always SHA256). So it may make more sense here to default to the curve's recommended hash.

Reply:
I don't understand this comment; this already defaults to curve's hash when `hash_algo` isn't defined. Are you saying we should disregard `hash_algo` and always use the curve's default hash? I'd assume this would break things.

Reply:
So, if you trace back to the upper-level functions, the hash algorithm is set here: https://github.com/openpgpjs/openpgpjs/blob/master/src/message.js#L371 - always to config.prefer_hash_algorithm which is set by default to SHA256. This made sense when we were just signing with RSA keys, but since different curves have different recommended hash algorithms I think we probably should use those instead.

Reply:
Although now that I think about it, we should be getting this value from the key if available rather than relying on the config.